# Prediction Change of Antibiotic Prescription 

# 배경
- Sepsis환자가 중환자실에 있는 동안 항생제를 유지할것인가? 바꿀것인가?


# 문제 종류
- binary classification: 항생제 처방을 바꿨는가 유지했는가

# 실험
- 코호트

<img src="images/pnp_fig1.png">

- 라벨링
    - TBD
<img src="images/ab-ex1.png">
<img src="images/ab-ex2.png">

- 피처 전처리/엔지니어링
    - TBD
- 모델
    - LR
    - RF
    - GB
    - (R)NN

# 교차검증
- 5-folds CV

# 평가
- AUROC
- variable importance: 어떤 변수가 다음 처방에 영향을 미쳤을까?


# 예측값의 활용
- 처방 실수의 예방

In [1]:
%matplotlib inline

import matplotlib.pyplot as plt
import seaborn as sns; sns.set(rc={'figure.figsize':(15,15)})
import numpy as np
import pandas as pd
from sqlalchemy import create_engine
engine = create_engine('postgresql://postgres:mimic@127.0.0.1:5555/mimic')

In [3]:
pd.read_sql("""
select * from abx_poe_list limit 10
""", engine)

drug  numobs
0                  Vancomycin   42633
1              Vancomycin HCl   22291
2                Levofloxacin   20867
3      MetRONIDAZOLE (FLagyl)    9985
4  Piperacillin-Tazobactam Na    9275
5                   CefazoLIN    8986
6                    CefePIME    8627
7     Piperacillin-Tazobactam    8482
8               Metronidazole    8302
9           Ciprofloxacin HCl    7559

In [23]:
pd.read_sql("""
select * from abx_micro_poe limit 10
""", engine)

icustay_id antibiotic_name antibiotic_time antibiotic_endtime  \
0      200001     CefTAZidime      2181-11-18         2181-11-20   
1      200001     CefTAZidime      2181-11-18         2181-11-20   
2      200001      Vancomycin      2181-11-19         2181-11-28   
3      200001      Vancomycin      2181-11-19         2181-11-28   
4      200001      Vancomycin      2181-11-19         2181-11-28   
5      200001      Vancomycin      2181-11-19         2181-11-19   
6      200001      Vancomycin      2181-11-19         2181-11-19   
7      200001      Vancomycin      2181-11-19         2181-11-19   
8      200001      Vancomycin      2181-11-19         2181-11-19   
9      200001      Vancomycin      2181-11-19         2181-11-19   

     last72_charttime    next24_charttime suspected_infection_time  \
0                 NaT 2181-11-18 11:45:00      2181-11-18 11:45:00   
1                 NaT 2181-11-18 11:10:00      2181-11-18 11:10:00   
2 2181-11-18 11:10:00 2181-11-19 06:00:00      2181-11-18 11:10:00   
3 2181-11-18 11:45:00 2181-11-19 04:18:00      2181-11-18 11:45:00   
4 2181-11-18 11:10:00 2181-11-19 04:18:00      2181-11-18 11:10:00   
5 2181-11-18 11:10:00 2181-11-19 13:00:00      2181-11-18 11:10:00   
6 2181-11-18 11:45:00 2181-11-19 13:00:00      2181-11-18 11:45:00   
7 2181-11-18 11:10:00 2181-11-19 04:18:00      2181-11-18 11:10:00   
8 2181-11-18 11:45:00 2181-11-19 04:18:00      2181-11-18 11:45:00   
9 2181-11-18 11:10:00 2181-11-19 06:00:00      2181-11-18 11:10:00   

        specimen  positiveculture  
0  BLOOD CULTURE                0  
1  BLOOD CULTURE                0  
2  BLOOD CULTURE                0  
3  BLOOD CULTURE                0  
4  BLOOD CULTURE                0  
5  BLOOD CULTURE                0  
6  BLOOD CULTURE                0  
7  BLOOD CULTURE                0  
8  BLOOD CULTURE                0  
9  BLOOD CULTURE                0

In [50]:
sql = """
select a.subject_id, s.* 
from sepsis3_cohort s 
    left join icustays a 
    on s.hadm_id=a.hadm_id and s.icustay_id=a.icustay_id
where s.excluded = 0
"""
label = pd.read_sql(sql, engine)
label.head().T

0  \
subject_id                                            165   
hadm_id                                            170252   
icustay_id                                         247247   
intime                                2170-10-03 17:17:53   
outtime                               2170-10-04 16:24:15   
age                                               86.4824   
gender                                                  M   
ethnicity                          BLACK/AFRICAN AMERICAN   
first_service                                         MED   
dbsource                                       metavision   
suspected_of_infection_poe                              1   
suspected_infection_time_poe          2170-10-03 11:20:00   
suspected_infection_time_poe_days                 0.24853   
specimen_poe                                BLOOD CULTURE   
positiveculture_poe                                     0   
antibiotic_time_poe                   2170-10-03 00:00:00   
exclusion_secondarystay                                 0   
exclusion_nonadult                                      0   
exclusion_csurg                                         0   
exclusion_carevue                                       0   
exclusion_early_suspicion                               0   
exclusion_late_suspicion                                0   
exclusion_bad_data                                      0   
excluded                                                0   

                                                        1  \
subject_id                                            266   
hadm_id                                            186251   
icustay_id                                         293876   
intime                                2168-07-10 08:02:23   
outtime                               2168-07-11 17:40:38   
age                                               77.5623   
gender                                                  F   
ethnicity                          BLACK/AFRICAN AMERICAN   
first_service                                        NMED   
dbsource                                       metavision   
suspected_of_infection_poe                              1   
suspected_infection_time_poe          2168-07-10 09:26:00   
suspected_infection_time_poe_days              -0.0580671   
specimen_poe                                  MRSA SCREEN   
positiveculture_poe                                     0   
antibiotic_time_poe                   2168-07-12 00:00:00   
exclusion_secondarystay                                 0   
exclusion_nonadult                                      0   
exclusion_csurg                                         0   
exclusion_carevue                                       0   
exclusion_early_suspicion                               0   
exclusion_late_suspicion                                0   
exclusion_bad_data                                      0   
excluded                                                0   

                                                     2                    3  \
subject_id                                         422                  671   
hadm_id                                         117029               126769   
icustay_id                                      299666               246119   
intime                             2173-04-03 11:36:29  2195-12-31 04:45:35   
outtime                            2173-04-08 14:20:31  2196-01-07 17:35:50   
age                                            57.6467              47.7629   
gender                                               M                    M   
ethnicity                                        WHITE                WHITE   
first_service                                     CMED                NSURG   
dbsource                                    metavision           metavision   
suspected_of_infection_poe                           0                    1   
suspected_infection_time_poe                       NaT  

In [29]:
label.groupby('excluded').size()

excluded
0    11791
1    49741
dtype: int64

In [30]:
label.subject_id.nunique()

46476

In [31]:
label[label.excluded==0].subject_id.nunique()

11791

In [32]:
label[label.excluded==1].subject_id.nunique()

36839

In [33]:
sql = """
select count(distinct subject_id)
from icustays
"""
pd.read_sql(sql, engine)

count
0  46476

- icu의 총 환자 46476명 중에 sepsis3 조건에 맞는 icustay는 11791 건
    - 해당 조건의 고유 환자수도 11791건으로 icustays:subejct_id = 1:1
    - 즉, 각 환자가 한번만 icu에 입원했고, sepsis 진단을 받았다. 여러번 입원하고 진단받은 사례는 없다.
    - 따라서, 환자 수준의 모델링만 가능, 방문 수준의 모델링은 불가능

<hr>

In [40]:
label[(label.excluded==0) & (label.antibiotic_time_poe.notnull())]['icustay_id'].count()

7588

In [41]:
label[(label.excluded==0) & (label.antibiotic_time_poe.isnull())]['icustay_id'].count()

4203

In [43]:
label[(label.excluded==0) & (label.antibiotic_time_poe.notnull())].subject_id.nunique()

7588

In [44]:
label[(label.excluded==0) & (label.antibiotic_time_poe.isnull())].subject_id.nunique()

4203

- icu의 총 환자 46476명 중에 sepsis3 조건에 맞는 icustay는 11791 건
    - 항생제 처방을 받은 환자수 7588, 건수 7588, 
    - 항생제 처방을 받지 않은 환자수 4203, 건수는 4203

<hr>

In [47]:
sql = """
select i.subject_id, s.*, a.antibiotic_time_poe
from sepsis3 s 
    join sepsis3_cohort a 
        on s.icustay_id=a.icustay_id
    join icustays i
        on s.icustay_id=i.icustay_id
where a.excluded = 0
"""
features = pd.read_sql(sql, engine)

In [48]:
features.head().T

0                    1  \
subject_id                                       61691                41710   
icustay_id                                      200021               200028   
hadm_id                                         109307               181955   
excluded                                             0                    0   
intime                             2114-12-26 19:45:12  2133-10-29 17:13:50   
outtime                            2114-12-27 22:46:28  2133-11-01 14:55:14   
dbsource                                    metavision           metavision   
suspected_infection_time_poe                       NaT                  NaT   
suspected_infection_time_poe_days                  NaN                  NaN   
specimen_poe                                      None                 None   
positiveculture_poe                                NaN                  NaN   
antibiotic_time_poe                                NaT                  NaT   
blood_culture_time                 2114-12-27 15:00:00  2133-10-29 19:27:00   
blood_culture_positive                               0                    0   
age                                            60.8501              64.8666   
gender                                               M                    M   
is_male                                              1                    1   
ethnicity                                        WHITE                WHITE   
race_white                                           1                    1   
race_black                                           0                    0   
race_hispanic                                        0                    0   
race_other                                           0                    0   
metastatic_cancer                                    0                    0   
diabetes                                             0                    1   
height                                             NaN               167.64   
weight                                            82.6                   84   
bmi                                                NaN              29.8899   
first_service                                     NMED                 CMED   
hospital_expire_flag                                 0                    0   
thirtyday_expire_flag                                0                    0   
icu_los                                         1.1259               2.9038   
hosp_los                                       1.94861              3.20417   
sepsis_angus                                         0                    0   
sepsis_martin                                        0                    0   
sepsis_explicit                                      0                    0   
septic_shock_explicit                                0                    0   
severe_sepsis_explicit                               0                    0   
sepsis_nqf                                           0                    0   
sepsis_cdc                                           0                    0   
sepsis_cdc_simple                                    0                    0   
elixhauser_hospital                                  4                    8   
vent                                                 1                    1   
sofa                                                 4                   10   
lods                                                 7                    6   
sirs                                                 3                    3   
qsofa                                                3                    2   
qsofa_sysbp_score                                    1                    1   
qsofa_gcs_score                                      1                    0   
qsofa_resprate_score                                 1                    1   
antibiotic_time_poe                                NaT                  NaT   

                       

# TBD

- sepsis3_cohort의 주요 피처들을 좀더 세밀한 타임 윈도우로 추출해야함
     - antibiotic_time_poe: 첫 처방인지 마지막 처방인지
     - vent 등 스코어나 체온 등을 세밀하게 뽑을 수 있는지 


In [5]:
df = pd.read_sql("""
with abx as
(
  select pr.hadm_id
  , pr.drug as antibiotic_name
  , pr.startdate as antibiotic_time
  , pr.enddate as antibiotic_endtime
  from prescriptions pr
  -- inner join to subselect to only antibiotic prescriptions
  inner join abx_poe_list ab
      on pr.drug = ab.drug
) select * from abx;
""", engine)

In [6]:
df.shape

(207154, 4)

In [7]:
df.head()

hadm_id           antibiotic_name antibiotic_time antibiotic_endtime
0   143045            Vancomycin HCl      2167-01-09         2167-01-11
1   159514         NEO*IV*Gentamicin      2117-11-20         2117-11-24
2   159514  NEO*IV*Ampicillin Sodium      2117-11-20         2117-11-24
3   150750              Levofloxacin      2149-11-10         2149-11-11
4   163353         NEO*IV*Gentamicin      2138-07-18         2138-07-20

In [8]:
ab_per_visit = df.groupby('hadm_id').agg('min')['antibiotic_time']

In [9]:
ab_per_visit.head()

hadm_id
100003   2150-04-17
100006   2108-04-06
100007   2145-03-31
100009   2162-05-17
100011   2177-08-29
Name: antibiotic_time, dtype: datetime64[ns]

In [24]:
df = pd.read_sql("""
select * from abx_micro_poe
""", engine)

In [25]:
df.shape

(1268126, 9)

In [26]:
df.head()

icustay_id antibiotic_name antibiotic_time antibiotic_endtime  \
0      200001     CefTAZidime      2181-11-18         2181-11-20   
1      200001     CefTAZidime      2181-11-18         2181-11-20   
2      200001      Vancomycin      2181-11-19         2181-11-28   
3      200001      Vancomycin      2181-11-19         2181-11-28   
4      200001      Vancomycin      2181-11-19         2181-11-28   

     last72_charttime    next24_charttime suspected_infection_time  \
0                 NaT 2181-11-18 11:45:00      2181-11-18 11:45:00   
1                 NaT 2181-11-18 11:10:00      2181-11-18 11:10:00   
2 2181-11-18 11:10:00 2181-11-19 06:00:00      2181-11-18 11:10:00   
3 2181-11-18 11:45:00 2181-11-19 04:18:00      2181-11-18 11:45:00   
4 2181-11-18 11:10:00 2181-11-19 04:18:00      2181-11-18 11:10:00   

        specimen  positiveculture  
0  BLOOD CULTURE              0.0  
1  BLOOD CULTURE              0.0  
2  BLOOD CULTURE              0.0  
3  BLOOD CULTURE              0.0  
4  BLOOD CULTURE              0.0

In [49]:
tmp = df[['icustay_id', 'antibiotic_name', 'positiveculture']].copy()
tmp['col'] = 1
target_abs = tmp.groupby(['icustay_id', 'positiveculture']).agg({'antibiotic_name': lambda x: tuple(set(x)), 'col': 'sum'})

target_abs.head()

antibiotic_name  \
icustay_id positiveculture                                                      
200001     0.0                                      (Vancomycin, CefTAZidime)   
           1.0                                                  (Vancomycin,)   
200003     0.0              (Sulfameth/Trimethoprim DS, *NF* Timentin, Van...   
           1.0              (Sulfameth/Trimethoprim DS, Vancomycin, Ciprof...   
200011     0.0                                   (Sulfameth/Trimethoprim DS,)   

                            col  
icustay_id positiveculture       
200001     0.0               77  
           1.0                7  
200003     0.0               65  
           1.0               22  
200011     0.0                1

In [56]:
icustay_id = 200003

In [61]:
pd.read_sql("select * from icustays where icustay_id=200003", engine)

row_id  subject_id  hadm_id  icustay_id dbsource first_careunit  \
0   34913       27513   163557      200003  carevue           SICU   

  last_careunit  first_wardid  last_wardid              intime  \
0          SICU            57           57 2199-08-02 19:50:04   

              outtime     los  
0 2199-08-08 17:09:18  5.8884

In [57]:
icustay1 = df[df.icustay_id == icustay_id].copy()

In [58]:
icustay1[icustay1.positiveculture == 1]

icustay_id             antibiotic_name antibiotic_time  \
89       200003  Piperacillin-Tazobactam Na      2199-08-02   
91       200003  Piperacillin-Tazobactam Na      2199-08-02   
92       200003                  Vancomycin      2199-08-02   
93       200003                  Vancomycin      2199-08-02   
96       200003                  Vancomycin      2199-08-04   
97       200003                  Vancomycin      2199-08-04   
100      200003               Ciprofloxacin      2199-08-04   
101      200003               Ciprofloxacin      2199-08-04   
102      200003                  Vancomycin      2199-08-04   
105      200003               Ciprofloxacin      2199-08-04   
106      200003               Ciprofloxacin      2199-08-04   
109      200003                  Vancomycin      2199-08-04   
111      200003                  Vancomycin      2199-08-06   
117      200003                  Vancomycin      2199-08-06   
123      200003        Ampicillin-Sulbactam      2199-08-08   
130      200003           Ciprofloxacin HCl      2199-08-08   
135      200003                  Vancomycin      2199-08-09   
139      200003                  Vancomycin      2199-08-09   
143      200003  Piperacillin-Tazobactam Na      2199-08-09   
166      200003                  Vancomycin      2199-08-18   
170      200003   Sulfameth/Trimethoprim DS      2199-08-22   
174      200003   Sulfameth/Trimethoprim DS      2199-08-22   

    antibiotic_endtime    last72_charttime    next24_charttime  \
89          2199-08-08                 NaT 2199-08-02 21:02:00   
91          2199-08-08                 NaT 2199-08-02 21:29:00   
92          2199-08-04                 NaT 2199-08-02 21:29:00   
93          2199-08-04                 NaT 2199-08-02 21:02:00   
96          2199-08-06 2199-08-02 21:02:00 2199-08-04 12:30:00   
97          2199-08-06 2199-08-02 21:29:00 2199-08-04 12:30:00   
100         2199-08-08 2199-08-02 21:02:00 2199-08-04 09:53:00   
101         2199-08-08 2199-08-02 21:29:00 2199-08-04 09:53:00   
102         2199-08-06 2199-08-02 21:02:00 2199-08-04 09:53:00   
105         2199-08-08 2199-08-02 21:02:00 2199-08-04 12:30:00   
106         2199-08-08 2199-08-02 21:29:00 2199-08-04 12:30:00   
109         2199-08-06 2199-08-02 21:29:00 2199-08-04 09:53:00   
111         2199-08-08 2199-08-04 09:53:00                 NaT   
117         2199-08-08 2199-08-05 11:06:00                 NaT   
123         2199-08-09 2199-08-05 11:06:00                 NaT   
130         2199-08-20 2199-08-05 11:06:00                 NaT   
135         2199-08-16                 NaT 2199-08-09 21:31:00   
139         2199-08-09                 NaT 2199-08-09 21:31:00   
143         2199-08-15                 NaT 2199-08-09 21:31:00   
166         2199-08-20 2199-08-17 01:27:00                 NaT   
170         2199-08-22 2199-08-20 19:58:00                 NaT   
174         2199-08-22 2199-08-20 20:00:00                 NaT   

    suspected_infection_time          specimen  positiveculture  
89       2199-08-02 21:02:00     BLOOD CULTURE              1.0  
91       2199-08-02 21:29:00     BLOOD CULTURE              1.0  
92       2199-08-02 21:29:00     BLOOD CULTURE              1.0  
93       2199-08-02 21:02:00     BLOOD CULTURE              1.0  
96       2199-08-02 21:02:00     BLOOD CULTURE              1.0  
97       2199-08-02 21:29:00     BLOOD CULTURE              1.0  
100      2199-08-02 21:02:00     BLOOD CULTURE              1.0  
101      2199-08-02 21:29:00     BLOOD CULTURE              1.0  
102      2199-08-02 21:02:00     BLOOD CULTURE              1.0  
105      2199-08-02 21:02:00     BLOOD CULTURE              1.0  
106      2199-08-02 21:29:00     BLOOD CULTURE              1.0  
109      2199-08-02 21:29:00     BLOOD CULTURE              1.0  
111      2199-08-04 09:53:00              BILE              1.0  
117      2199-08-05 11:06:00              BILE              1.0  
123      2199-08-05 11:06:00              BILE

In [53]:
icustay1.antibiotic_name.nunique(), icustay1.antibiotic_name.unique()

(8, array(['Vancomycin', 'Piperacillin-Tazobactam Na', 'Ciprofloxacin',
        'Ampicillin-Sulbactam', 'Ciprofloxacin HCl',
        'Sulfameth/Trimethoprim', '*NF* Timentin',
        'Sulfameth/Trimethoprim DS'], dtype=object))

In [54]:
icustay1.groupby('antibiotic_name').agg(min)['antibiotic_time']

antibiotic_name
*NF* Timentin                2199-08-16
Ampicillin-Sulbactam         2199-08-08
Ciprofloxacin                2199-08-04
Ciprofloxacin HCl            2199-08-08
Piperacillin-Tazobactam Na   2199-08-02
Sulfameth/Trimethoprim       2199-08-16
Sulfameth/Trimethoprim DS    2199-08-22
Vancomycin                   2199-08-02
Name: antibiotic_time, dtype: datetime64[ns]

In [55]:
icustay1.groupby('antibiotic_name').agg(max)['antibiotic_endtime']

antibiotic_name
*NF* Timentin                2199-08-16
Ampicillin-Sulbactam         2199-08-09
Ciprofloxacin                2199-08-08
Ciprofloxacin HCl            2199-08-20
Piperacillin-Tazobactam Na   2199-08-15
Sulfameth/Trimethoprim       2199-08-22
Sulfameth/Trimethoprim DS    2199-08-22
Vancomycin                   2199-08-20
Name: antibiotic_endtime, dtype: datetime64[ns]

# Reference 

- [1] Sepsis-3 in MIMIC-III, https://github.com/alistairewj/sepsis3-mimic
- [2] Antibiotic prescription of Sepsis-3 in MIMIC-III, https://github.com/alistairewj/sepsis3-mimic/blob/master/query/tbls/abx-micro-prescription.sql